### PR TITLE
Implement TH_API functions from the set 4

### PR DESCRIPTION
**Pull request by VirrageS**

### CODE REVIEW

**Comment by jytug** [comment]
Hey, I've tried to compile your branch on a machine with a proper GPU and CUDA support, but it failed. [Here](http://termbin.com/23ur) is the whole log of `python setup.py build_deps`.
I'll take a look at the code in a while (i.e. tomorrow)

**Comment by VirrageS** [comment]
I have fixed some of these issues. Not sure if all of them.

File: torch/lib/THPP/generators/THCGenerator.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#include "THCGenerator.hpp"
+
+#define const_generator_cast(generator) \
+  dynamic_cast<const THCGenerator&>(generator)
+
+namespace thpp {
+
+THCGenerator::THCGenerator(THCState* state))
```
Comment (jytug) [inline torch/lib/THPP/generators/THCGenerator.cpp]: Won't compile :-)

File: torch/lib/THPP/generators/THCGenerator.hpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+};
+
+} // namespace thpp
+
+#include "../Generator.hpp"
+
+namespace thpp {
+
+template<typename real>
+struct THCGenerator : public Generator {
```
Comment (jytug) [inline torch/lib/THPP/generators/THCGenerator.hpp]: Is it really a template?

File: torch/lib/THPP/generators/THCGenerator.cpp (hunk 1 above)
Comment (jytug) [inline torch/lib/THPP/generators/THCGenerator.cpp]: Won't compile ;-)

File: torch/lib/THPP/generators/THCGenerator.hpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  using generator_type = Generator;
+};
+
+} // namespace thpp
+
+#include "../Generator.hpp"
+
+namespace thpp {
+
+template<typename real>
```
Comment (jytug) [inline torch/lib/THPP/generators/THCGenerator.hpp]: I don't think it's a template

File: torch/lib/THPP/generators/THCGenerator.hpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+struct THCGenerator : public Generator {
+  using generator_type = typename thc_generator_traits::generator_type;
+
+  THCGenerator(THCState* state);
+  virtual ~THCGenerator();
+
+  virtual THCGenerator& copy(const Generator& from) override;
+  virtual THCGenerator& free() override;
+
+  virtual THCGenerator& seed() override;
```
Comment (jytug) [inline torch/lib/THPP/generators/THCGenerator.hpp]: It should return `unsigned long`

File: torch/lib/THPP/generators/THCGenerator.cpp [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+  : generator(THCRandom_getGenerator())
```
Comment (jytug) [inline torch/lib/THPP/generators/THCGenerator.cpp]: `THCRandom_getGenerator()` is not declared in any `.h` file, so this won't compile either - moreover, the `THCRandom_getGenerator` from `THC` takes a `state` argument

File: torch/lib/THPP/generators/THCGenerator.cpp [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  cudaGetDevice(&current_device);
+  THCRandom_init(state, num_devices, current_device);
+}
+
+THCGenerator::~THCGenerator() {
+  THCRandom_shutdown(state);
+}
+
+THCGenerator& THCGenerator::copy(const Generator& from) {
+  throw std::runtime_error("THCGenerator::copy() not implemented");
```
Comment (jytug) [inline torch/lib/THPP/generators/THCGenerator.cpp:24]: Include the `stdexcept` header

**Review by jytug — CHANGES_REQUESTED** [review]

File: torch/lib/THD/master_worker/master/THDRandom.cpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+#include <cstring>
+
+using namespace thd;
+using namespace rpc;
+using namespace master;
+
+static THDGenerator* THDGenerator_newUnseeded() {
+  THDGenerator *new_generator = new THDGenerator();
+  std::memset(reinterpret_cast<void*>(new_generator), 0, sizeof(new_generator));
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/THDRandom.cpp]: Why is that even needed? You're setting the only field below. If you want to ensure this is going to be correct in the future, make the only constructor accept an id.

File: torch/lib/THD/master_worker/master/THDRandom.cpp [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  delete self;
+}
+
+unsigned long THDRandom_seed(THDGenerator *_generator) {
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::generatorSeed, _generator),
+    THDState::s_current_worker
+  );
+
+  return receiveValueFromWorker<unsigned long>(THDState::s_current_worker);
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/THDRandom.cpp:58]: What's that value? The used seed?

File: torch/lib/THD/master_worker/master/THDRandom.cpp [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+void THDRandom_manualSeed(THDGenerator *_generator, unsigned long the_seed_) {
+  THDGenerator *blank = THDGenerator_newUnseeded();
+  THDGenerator_copy(_generator, blank);
+  THDGenerator_free(blank);
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/THDRandom.cpp]: I don't get that part. Why is that needed?

File: torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp [hunk 10]
```diff
@@ -302,8 +329,271 @@ TENSOR_IMPLEMENT_POINTWISE_FUNCTION(cosh,Cosh)
...
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorLogspace, r_, n, a, b),
+    THDState::s_current_worker
+  );
+}
+
+void THDTensor_(rand)(THDTensor *r_, THDGenerator *_generator,
+                      THLongStorage *size) {
+  THDTensor_(resize)(r_, size, NULL);
+  THDTensor_(uniform)(r_, _generator, 0, 1);
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp]: Why is that even needed? Either call resize + uniform, or resize and send a message.

File: torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp [hunk 11]
```diff
@@ -302,8 +329,271 @@ TENSOR_IMPLEMENT_POINTWISE_FUNCTION(cosh,Cosh)
...
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorRandn, r_, _generator, size),
+    THDState::s_current_worker
+  );
+}
+
+void THDTensor_(histc)(THDTensor *hist, THDTensor *tensor, long nbins,
+                       real minvalue, real maxvalue) {
+  THDTensor_(resize1d)(hist, nbins);
+  THDTensor_(zero)(hist);
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp]: I don't think we need that.

File: torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp [hunk 12]
```diff
@@ -302,8 +329,271 @@ TENSOR_IMPLEMENT_POINTWISE_FUNCTION(cosh,Cosh)
...
+  masterCommandChannel->sendMessage(
+    packMessage(Functions::tensorRand, r_, _generator, size),
+    THDState::s_current_worker
+  );
+}
+
+void THDTensor_(randn)(THDTensor *r_, THDGenerator *_generator,
+                       THLongStorage *size) {
+  THDTensor_(resize)(r_, size, NULL);
+  THDTensor_(normal)(r_, _generator, 0, 1);
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp]: That seems unnecessary

File: torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp [hunk 13]
```diff
@@ -302,8 +329,271 @@ TENSOR_IMPLEMENT_POINTWISE_FUNCTION(cosh,Cosh)
...
+  THArgCheck(THDTensor_(nDimension)(tensor) < 3, 2,
+             "invalid dimension %d, the input must be a 2d tensor",
+             THDTensor_(nDimension)(tensor));
+
+  int dimension = 1;
+  THArgCheck(dimension >= 0 && dimension < THDTensor_(nDimension)(tensor), 2,
+             "invalid dimension %d", dimension + TH_INDEX_BASE);
+
+  THDTensor_(resize2d)(hist, tensor->size[0], nbins);
+  THDTensor_(zero)(hist);
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/generic/THDTensorMath.cpp]: Remove this.

File: torch/lib/THD/master_worker/master/generic/THDTensorRandom.cpp [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+  if (!with_replacement) {
+    THArgCheck((!with_replacement) && (n_sample <= n_categories), 2, \
+    "cannot sample n_sample > prob_dist:size(1) samples without replacement");
+  }
+
+  /* will contain multinomial samples (category indices to be returned) */
+  THDLongTensor_resize2d(self, n_dist , n_sample);
+
+  for (int i = 0; i < n_dist; i++) {
```
Comment (apaszke) [inline torch/lib/THD/master_worker/master/generic/THDTensorRandom.cpp]: I don't think we need to do this kind of error checking in the master. It's quite unlikely and may have a performance impact.

File: torch/lib/THD/master_worker/worker/dispatch/Tensor.cpp [hunk 15]
```diff
@@ -354,20 +354,6 @@ static void tensorProdall(rpc::RPCMessage& raw_message) {
   }
 }
 
-static void tensorNeg(rpc::RPCMessage& raw_message) {
```
Comment (apaszke) [inline torch/lib/THD/master_worker/worker/dispatch/Tensor.cpp:357]: Why did you remove these?

File: torch/lib/THPP/tensors/generic/THCTensor.cpp [hunk 16]
```diff
@@ -542,6 +551,484 @@ auto THCTensor<real>::sinh(const Tensor& src) -> THCTensor& {
...
+      floating point tensors");
+#endif
+}
+
+template<>
+int THCTensor<real>::logicalany() {
+#if defined(TH_REAL_IS_BYTE)
+  return THCTensor_(logicalany)(state, tensor);
+#else
+  throw std::runtime_error("floating point functions are available only for\
```
Comment (apaszke) [inline torch/lib/THPP/tensors/generic/THCTensor.cpp]: These are byte tensor functions.

**Review by apaszke — COMMENTED** [review]
Looks very good for the most part. Fix the comments and should be good to go

File: torch/lib/THD/master_worker/master/THDRandom.cpp (hunk 8 above)
Comment (VirrageS) [reply torch/lib/THD/master_worker/master/THDRandom.cpp:58]: Yes, `seed` function returns the used seed.

File: torch/lib/THD/master_worker/master/THDRandom.cpp (hunk 9 above)
Comment (VirrageS) [reply torch/lib/THD/master_worker/master/THDRandom.cpp]: It is unnecessary I forgot to delete this. Previously I had more fields in `THDGenerator` struct and this was to clear state of generator I think. From TH:
```
  /* This ensures reseeding resets all of the state (i.e. state for Gaussian numbers) */
```

File: torch/lib/THD/master_worker/worker/dispatch/Tensor.cpp (hunk 15 above)
Comment (VirrageS) [reply torch/lib/THD/master_worker/worker/dispatch/Tensor.cpp:357]: I just moved it to another file: https://github.com/apaszke/pytorch-dist/pull/58/files#diff-c0e7ff130c09526e4236f3b315eecf23R344.

Same with `cinv`.

**Review by apaszke — APPROVED** [review]

**Pull request merged**